IRGen: Don't eagerly initialize NSCoding conformers on startup on new deployment targets

slavapestov opened the pull request:
If objc_getClass() can realize arbitrary Swift classes there's no reason
to generate the static constructor.

Fixes <rdar://problem/49660515>.

slavapestov commented:
@jrose-apple Before I update the tests and so on -- does this make sense to you?

jrose-apple commented:
Eagerly registering a class also runs any `+initialize` it may have, so I'd prefer if you put the check in Sema where the attribute gets added instead. That said, we currently ban writing `+initialize` implementations in Swift, and don't currently synthesize any, so this will have the same effect in practice today.

slavapestov commented:
@jrose-apple I understand where you're coming from but as far as I can tell the only effect of the attribute is to add the class to the vector in IRGen. IRGen emits a static constructor, which calls the metadata accessor for each eager class which I assume triggers the call to `+initialize`. So I don't think my change would be any different than not adding the attribute in Sema in the first place. Am I missing something? Regardless I can do it in Sema too, I was mostly just being lazy because I already have the right deployment check factored out in IRGen.

slavapestov commented:
@jrose-apple I went and looked and it turns out we have two attributes:

- `StaticInitializeObjCMetadataAttr` is the one I'm changing in this PR, and it only determines if we emit the global constructor or not.
- `ObjCNonLazyRealizationAttr` this one is different. It can only be used for classes with fixed metadata (no generic ancestry) and as far as I see it only exists so that the empty array and empty dictionary singletons can be statically allocated with valid isa pointers in the stdlib. This is probably the one you're thinking of since it triggers Objective-C realization on startup without an opportunity to run Swift metadata accessors.

jrose-apple commented:
I was thinking that we may have a use for the `StaticInitializeObjCMetadataAttr` case that isn't just "make sure it can be found by NSClassFromString". We don't today, but the name of the attribute implies that we might. Maybe renaming the attribute would be another way to go?

jrose-apple commented:
The two attributes have basically the same effect for a class with fixed metadata; it's just that the latter is something the Objective-C runtime handles natively and the former uses a static constructor.

slavapestov commented:
@jrose-apple Right. We don't actually ever emit StaticInitializeObjCMetadataAddr on classes without generic ancestry, since those could be found by name on the old runtime anyway. I'm not proposing renaming or removing the attribute, just not adding it on classes conforming to NSCoding when your deployment target is sufficiently new.

jrose-apple commented:
I'm in favor of "don't add the attribute on classes conforming to NSCoding", but this patch doesn't do that; it ignores the attribute in IRGen instead.

slavapestov commented:
@jrose-apple Ok, glad we're on the same page then.

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on include/swift/Basic/LangOptions.h:
```diff
@@ -371,6 +371,24 @@ namespace swift {
+    // - iOS 12.2
+    // - tvOS 12.2
+    // - watchOS 5.2
+    bool doesPlatformSupportObjCGetClassHook() const;
```
Nitpick: these should be "target" or "deployment target" or "minimum deployment target", not "platform". "Platform" doesn't include a version number.

jrose-apple commented on lib/Basic/LangOptions.cpp:
```diff
@@ -275,3 +275,18 @@ std::pair<bool, bool> LangOptions::setTarget(llvm::Triple triple) {
+  if (Target.isWatchOS())
+    return !Target.isOSVersionLT(5, 2);
+
+  return false;
```
Can you leave a comment noting that this allows us to run ObjC tests on non-Apple platforms? That's the main reason to leave it this way instead of asserting, I think.

jrose-apple commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -5039,7 +5043,8 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
           isNSCoding(conformance->getProtocol()) &&
-          !classDecl->isGenericContext()) {
+          !classDecl->isGenericContext() &&
+          !classDecl->hasClangNode()) {
```
When does this come up?

slavapestov replied:
Retroactive conformance of imported class to NSCoding. We already checked for hasClangNode() before, but I moved it.

jrose-apple replied:
\*shrug\* All right. It's true that those won't have unstable names either.

jrose-apple commented on test/Interpreter/SDK/archive_attributes.swift:
```diff
@@ -1,7 +1,6 @@
-// RUN: %target-build-swift %s -module-name=test -Xfrontend -disable-llvm-optzns -emit-ir | %FileCheck -check-prefix=CHECK-IR %s
 // RUN: %target-run %t/encode %t/test.arc
 // RUN: plutil -p %t/test.arc | %FileCheck -check-prefix=CHECK-ARCHIVE %s
 // RUN: %target-run %t/decode %t/test.arc | %FileCheck %s
```
This is an execution test; would it make sense to do it both the old way and the new way if we're on a new enough OS?

jrose-apple replied:
(I'm not even 100% sure how to write such a test, but…)

slavapestov replied:
I can build the test with an old and new deployment target, and in the new deployment target case, add a runtime check that skips the test.

slavapestov committed:
IRGen: Move some deployment target checks to LangOptions

slavapestov committed:
IRGen: Disable eager initialization of NSCoding adopters on newer targets

If a class does not have a custom @objc name, objc_getClass() can find
it at runtime by calling the Swift runtime's metadata demangler hook.

This avoids the static initializer on startup. If the class has a
custom runtime name we still need the static initializer unfortunately.

Fixes <rdar://problem/49660515>.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test macOS
